src/canister_tests/src/tests.rs

```diff
@@ -66,6 +66,462 @@ fn registration_with_mismatched_sender_fails() {
+                .duration_since(UNIX_EPOCH)
+                .unwrap()
+                .as_nanos() as u64
+        );
+
+        let signed_delegation = match api::get_delegation(
+            &env,
+            canister_id,
+            principal_1(),
+            user_number,
+            frontend_hostname.to_string(),
+            pub_session_key.clone(),
+            expiration,
+        )? {
+            GetDelegationResponse::SignedDelegation(delegation) => delegation,
+            GetDelegationResponse::NoSuchDelegation => panic!("failed to get delegation"),
+        };
+
+        framework::verify_delegation(&env, canister_sig_key, &signed_delegation);
+        assert_eq!(signed_delegation.delegation.pubkey, pub_session_key);
+        assert_eq!(signed_delegation.delegation.expiration, expiration);
+        Ok(())
+    }
+
+    /// Verifies that delegations can be requested in parallel.
+    #[test]
+    fn should_get_multiple_valid_delegations() -> Result<(), CallError> {
+        let env = StateMachine::new();
+        let canister_id = framework::install_ii_canister(&env, framework::II_WASM.clone());
+        let user_number = flows::register_anchor(&env, canister_id);
+        let frontend_hostname_1 = "https://dapp1.com";
+        let frontend_hostname_2 = "https://dapp2.com";
+        let pub_session_key_1 = ByteBuf::from("session public key 1");
+        let pub_session_key_2 = ByteBuf::from("session public key 2");
+        let delegation_params = vec![
+            (
+                &pub_session_key_1,
+                frontend_hostname_1,
+                SystemTime::UNIX_EPOCH,
+            ),
+            (
+                &pub_session_key_1,
+                frontend_hostname_2,
+                SystemTime::UNIX_EPOCH,
+            ),
+            (
+                &pub_session_key_2,
+                frontend_hostname_1,
+                SystemTime::UNIX_EPOCH,
+            ),
+            (
+                &pub_session_key_1,
+                frontend_hostname_1,
+                SystemTime::UNIX_EPOCH.add(Duration::from_secs(30)),
+            ),
+        ];
+
+        let mut prepare_delegation_results = vec![];
+        for (session_key, frontend_hostname, time) in delegation_params.clone() {
+            env.set_time(time);
+            let (canister_sig_key, expiration) = api::prepare_delegation(
+                &env,
+                canister_id,
+                principal_1(),
+                user_number,
+                frontend_hostname.to_string(),
+                session_key.clone(),
+                None,
+            )?;
+
+            assert_eq!(
+                expiration,
+                env.time()
+                    .add(Duration::from_secs(30 * 60)) // default expiration: 30 minutes
+                    .duration_since(UNIX_EPOCH)
+                    .unwrap()
+                    .as_nanos() as u64
+            );
+            prepare_delegation_results.push((canister_sig_key, expiration))
```

Comment:
why create `prepare_delegation_results` and not simply check the delegation here?

Reply:
ah I see! then you can probably get rid of the intermediate structures + get rid of a lot of room for mishandling indices by doing something like this:

``` rust
        let foo = delegation_params.clone().into_iter().map(|(session_key, frontend_hostname, time)| {
            env.set_time(time);
            let (canister_sig_key, expiration) = api::prepare_delegation(
                &env,
                canister_id,
                principal_1(),
                user_number,
                frontend_hostname.to_string(),
                session_key.clone(),
                None,
            ).unwrap();

            assert_eq!(
                expiration,
                env.time()
                    .add(Duration::from_secs(30 * 60)) // default expiration: 30 minutes
                    .duration_since(UNIX_EPOCH)
                    .unwrap()
                    .as_nanos() as u64
            );
            (session_key, frontend_hostname, canister_sig_key, expiration)
        });

        for (session_key, frontend_hostname, canister_sig_key, expiration) in foo {
            let signed_delegation = match api::get_delegation(
                &env,
                canister_id,
                principal_1(),
                user_number,
                frontend_hostname.to_string(),
                (*session_key).clone(),
                expiration,
            )? {
                GetDelegationResponse::SignedDelegation(delegation) => delegation,
                GetDelegationResponse::NoSuchDelegation => panic!("failed to get delegation"),
            };

            framework::verify_delegation(&env, canister_sig_key.clone(), &signed_delegation);
            assert_eq!(signed_delegation.delegation.pubkey, (*session_key).clone());
            assert_eq!(signed_delegation.delegation.expiration, expiration);
        };
```